### PR TITLE
Improve the TOTP

### PR DESCRIPTION
**Pull request by Kanapriya**

### CODE REVIEW

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPAuthenticatorKey.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+public class TOTPAuthenticatorKey {
+    /**
+     * The format string to generate the URL of a Google-provided QR bar code.
+     *
+     * @deprecated Use GoogleAuthenticatorQRGenerator instead.
+     */
+    private static final String QR_FORMAT =
```
Comment (keerthu) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPAuthenticatorKey.java]: I think we don't need this. So remove this variable.

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPAuthenticatorKey.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+     * application on their smart phones or manually enter the secret manually.
+     *
+     * @param user   the user to assign the secret key to.
+     * @param host   the host to assign the secret key to.
+     * @param secret the secret key in Base32 encoding.
+     * @return the URL of a Google-provided QR barcode to be loaded into the
+     * Google Authenticator application.
+     * @deprecated Use GoogleAuthenticatorQRGenerator##getOtpAuthURL instead.
+     */
+    @SuppressWarnings("deprecation, unused")
```
Comment (keerthu) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPAuthenticatorKey.java]: Remove this method.

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPAuthenticatorImpl.java [hunk 3]
```diff
@@ -0,0 +1,447 @@
...
+     * This method calculates the validation code at time 0.
+     *
+     * @param secretKey The secret key to use.
+     * @return the validation code at time 0.
+     */
+    private int calculateValidationCode(byte[] secretKey) {
+        return calculateCode(secretKey, 0);
+    }
+
+
```
Comment (keerthu) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPAuthenticatorImpl.java]: Remove space.

File: component/authenticator/pom.xml [hunk 4]
```diff
@@ -1,112 +1,108 @@
...
+                    <instructions>
+                        <Bundle-SymbolicName>${project.artifactId}</Bundle-SymbolicName>
+                        <Bundle-Name>${project.artifactId}</Bundle-Name>
+                        <Import-Package>
+                            org.wso2.carbon.core.utils.*,
+                            org.wso2.carbon.core.*,
+                            org.osgi.service.*,
+                            org.wso2.carbon.registry.core.*,
+                            org.wso2.carbon.utils.*,
+                            org.apache.commons.logging.*;version="${commons-logging.osgi.version.range}",
```
Comment (Kanapriya) [inline component/authenticator/pom.xml]: From parent pom

File: component/authenticator/pom.xml [hunk 5]
```diff
@@ -1,112 +1,108 @@
...
+                            org.osgi.framework.*; version="${osgi.framework.imp.pkg.version.range}",
```
Comment (Kanapriya) [inline component/authenticator/pom.xml]: From parent pom

File: component/authenticator/pom.xml [hunk 6]
```diff
@@ -1,112 +1,108 @@
...
+                            org.wso2.carbon.identity.application.authentication.framework.*;version="${carbon.identity.package.import.version.range}",
```
Comment (Kanapriya) [inline component/authenticator/pom.xml]: From parent pom

File: component/authenticator/pom.xml [hunk 7]
```diff
@@ -1,112 +1,108 @@
...
+                            javax.servlet;version="${imp.pkg.version.javax.servlet}",
```
Comment (Kanapriya) [inline component/authenticator/pom.xml]: From parent pom

File: component/authenticator/pom.xml [hunk 8]
```diff
@@ -1,112 +1,108 @@
...
+                            javax.servlet.http;version="${imp.pkg.version.javax.servlet}",
```
Comment (Kanapriya) [inline component/authenticator/pom.xml]: From parent pom

File: component/authenticator/pom.xml [hunk 9]
```diff
@@ -1,112 +1,108 @@
...
+                            org.wso2.carbon.identity.core.*; version="${carbon.identity.package.import.version.range}",
```
Comment (Kanapriya) [inline component/authenticator/pom.xml]: From parent pom

File: component/authenticator/pom.xml [hunk 10]
```diff
@@ -1,112 +1,108 @@
...
+                            org.wso2.carbon.identity.mgt.*; version="${carbon.identity.package.import.version.range}"
```
Comment (Kanapriya) [inline component/authenticator/pom.xml]: From parent pom

File: component/authenticator/pom.xml [hunk 11]
```diff
@@ -1,112 +1,108 @@
...
+            </plugin>
+            <plugin>
+                <groupId>org.apache.felix</groupId>
+                <artifactId>maven-bundle-plugin</artifactId>
+                <extensions>true</extensions>
+                <configuration>
+                    <instructions>
+                        <Bundle-SymbolicName>${project.artifactId}</Bundle-SymbolicName>
+                        <Bundle-Name>${project.artifactId}</Bundle-Name>
+                        <Import-Package>
```
Comment (Kanapriya) [inline component/authenticator/pom.xml]: Fixed in https://github.com/wso2-extensions/identity-outbound-auth-totp/pull/16

File: component/authenticator/pom.xml [hunk 12]
```diff
@@ -1,112 +1,108 @@
...
+                            org.osgi.framework.*; version="${osgi.framework.imp.pkg.version.range}",
+                            org.wso2.carbon.identity.application.authentication.framework.*;version="${carbon.identity.package.import.version.range}",
+                            javax.servlet;version="${imp.pkg.version.javax.servlet}",
+                            javax.servlet.http;version="${imp.pkg.version.javax.servlet}",
+                            org.wso2.carbon.identity.core.*; version="${carbon.identity.package.import.version.range}",
+                            org.wso2.carbon.identity.mgt.*; version="${carbon.identity.package.import.version.range}"
+                        </Import-Package>
+                        <DynamicImport-Package>*</DynamicImport-Package>
+                        <Export-Package>!org.wso2.carbon.identity.application.authenticator.totp.internal.*,
+                            org.wso2.carbon.identity.application.authenticator.totp.*
```
Comment (Kanapriya) [inline component/authenticator/pom.xml]: fixed in https://github.com/wso2-extensions/identity-outbound-auth-totp/pull/16

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPTokenVerifier.java [hunk 13]
```diff
@@ -72,22 +72,22 @@ public static TOTPTokenVerifier getInstance() {
...
         if (TOTPAuthenticatorConstants.BASE64.equals(TOTPUtil.getEncodingMethod())) {
-            encoding = KeyRepresentation.BASE64;
+            encoding = TOTPKeyRepresentation.BASE64;
         }
         timeStep = TimeUnit.SECONDS.toMillis(TOTPUtil.getTimeStepSize());
         windowSize = TOTPUtil.getWindowSize();
 
-        GoogleAuthenticatorConfig.GoogleAuthenticatorConfigBuilder gacb = new GoogleAuthenticatorConfig
-                .GoogleAuthenticatorConfigBuilder()
+        TOTPAuthenticatorConfig.TOTPAuthenticatorConfigBuilder gacb = new TOTPAuthenticatorConfig
```
Comment (Kanapriya) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPTokenVerifier.java]: Fixed in https://github.com/wso2-extensions/identity-outbound-auth-totp/pull/16

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPAuthenticatorConfig.java [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.application.authenticator.totp.util;
+
+import java.util.concurrent.TimeUnit;
+
```
Comment (Kanapriya) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPAuthenticatorConfig.java]: Fixed in https://github.com/wso2-extensions/identity-outbound-auth-totp/pull/16

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPAuthenticatorConfig.java [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+        return windowSize;
+    }
+
+    public static class TOTPAuthenticatorConfigBuilder {
+        private TOTPAuthenticatorConfig config = new TOTPAuthenticatorConfig();
+
+        public TOTPAuthenticatorConfig build() {
+            return config;
+        }
+
```
Comment (Kanapriya) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPAuthenticatorConfig.java]: Fixed in https://github.com/wso2-extensions/identity-outbound-auth-totp/pull/16

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPReseedingSecureRandom.java [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+ */
+
+package org.wso2.carbon.identity.application.authenticator.totp.util;
+
+import java.security.NoSuchAlgorithmException;
+import java.security.NoSuchProviderException;
+import java.security.SecureRandom;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class TOTPReseedingSecureRandom {
```
Comment (hmrajas) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPReseedingSecureRandom.java]: Please add class level comments and method level comments to all class and methods.

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPReseedingSecureRandom.java [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+        if (provider == null) {
+            throw new IllegalArgumentException("Provider cannot be null.");
+        }
+
+        this.algorithm = algorithm;
+        this.provider = provider;
+
+        buildSecureRandom();
+    }
+
```
Comment (hmrajas) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPReseedingSecureRandom.java]: Please add class level comments and method level comments to all class and methods.

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPReseedingSecureRandom.java [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+    @SuppressWarnings("UnusedDeclaration")
+    TOTPReseedingSecureRandom() {
+        this.algorithm = null;
+        this.provider = null;
+        buildSecureRandom();
+    }
+
+    @SuppressWarnings("UnusedDeclaration")
+    TOTPReseedingSecureRandom(String algorithm) {
+        if (algorithm == null) {
```
Comment (keerthu) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPReseedingSecureRandom.java]: Use StringUtils.isEmpty() for null check.

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPReseedingSecureRandom.java (hunk 16 above)
Comment (Kanapriya) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPReseedingSecureRandom.java]: Fixed in 82c1b2dad7d0907f39b0bc8b227a14ffedf56820

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPReseedingSecureRandom.java (hunk 17 above)
Comment (Kanapriya) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPReseedingSecureRandom.java]: Fixed in 82c1b2dad7d0907f39b0bc8b227a14ffedf56820

File: component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPReseedingSecureRandom.java (hunk 18 above)
Comment (Kanapriya) [inline component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/util/TOTPReseedingSecureRandom.java]: Fixed in 82c1b2dad7d0907f39b0bc8b227a14ffedf56820

**Pull request merged**